ink-wrapper/src/extensions.rs

```diff
@@ -0,0 +1,131 @@
+use ink_metadata::MessageSpec;
+use scale_info::{form::PortableForm, Field, Type, TypeDefComposite, Variant};
+
+pub trait TypeExtensions {
+    /// Returns true if the type is a rust primitive.
+    fn is_primitive(&self) -> bool;
+
+    /// Returns true if the type is defined in the ink_primitives crate.
+    fn is_ink(&self) -> bool;
+
+    /// Returns true if the type is defined in the ink_primitives crate's private types module. These types are
+    /// reexported at the top level in ink_primitives.
+    fn is_ink_types(&self) -> bool;
+
+    /// Returns true if the type is a builtin type.
+    fn is_builtin(&self) -> bool;
+
+    /// Returns the name by which the type can be referenced.
+    ///
+    /// It's the full path to the type for ink! types and just the name for other types. That's because any custom types
+    /// for the contract will be defined in the same module as the functions that use them.
+    fn qualified_name(&self) -> String;
+}
+
+impl TypeExtensions for Type<PortableForm> {
+    fn is_primitive(&self) -> bool {
+        matches!(self.type_def(), scale_info::TypeDef::Primitive(_))
+    }
+
+    fn is_ink(&self) -> bool {
+        !self.path().segments().is_empty() && self.path().segments()[0] == "ink_primitives"
+    }
+
+    fn is_ink_types(&self) -> bool {
+        self.path().segments().len() > 2
+            && self.path().segments()[0] == "ink_primitives"
+            && self.path().segments()[1] == "types"
+    }
+
+    fn is_builtin(&self) -> bool {
+        self.path().segments().len() == 1
+    }
+
+    fn qualified_name(&self) -> String {
+        if self.is_ink_types() {
+            ["ink_primitives", self.path().segments().last().unwrap()].join("::")
+        } else if self.is_ink() {
+            self.path().segments().join("::")
+        } else {
+            self.path().segments().last().unwrap().to_string()
+        }
+    }
+}
+
+pub trait MessageSpecExtensions {
+    fn trait_name(&self) -> Option<String>;
+    fn method_name(&self) -> String;
+}
+
+impl MessageSpecExtensions for MessageSpec<PortableForm> {
+    fn trait_name(&self) -> Option<String> {
+        let parts = self.label().split("::").collect::<Vec<&str>>();
+        if parts.len() == 2 {
+            Some(parts[0].to_string())
+        } else {
+            None
+        }
```

Comment:
Why more than 2 parts in the path are `None`? i.e. `openbrush::psp22::PSP22` vs `psp22::PSP22`?

Reply:
So far the only cases are `regular_contract_method_name` and `PSP22::transfer` or somesuch, so this is what I went with. If anything, we could `panic!` on length other than 1 or 2, because we don't know what to do in such cases.

Reply:
:+1: